homeassistant/components/mqtt/mixins.py

```diff
@@ -727,6 +739,16 @@ async def discovery_callback(payload):
                     _LOGGER.info("Ignoring unchanged update for: %s", self.entity_id)
             send_discovery_done(self.hass, self._discovery_data)
 
+        @callback
+        def _entry_unload(*_: Any) -> None:
+            """Handle the unload of the config entry."""
+            _LOGGER.info("Removing component: %s", self.entity_id)
+            self._cleanup_discovery_on_remove()
+            # cleanup platform resources
+            self.hass.async_create_task(_async_remove_state_and_registry_entry(self))
```

Comment:
Why do we want to remove the entity from the registries?

Reply:
According to https://developers.home-assistant.io/docs/config_entries_index?_highlight=async_unload_entry#unloading-entries we already do the cleanup on the right spot. But when `async_unload_entry(hass, entry)` is called the cleanup is not finished yet.